### PR TITLE
Graph printer

### PR DESCRIPTION
**Pull request by jenrik**
The graph printer will produce a graphviz graph of the abstract syntax
tree.
This PR depends on changes in #62

### CODE REVIEW

File: compiler/src/main/kotlin/walkers/grapher.kt [hunk 1]
```diff
@@ -0,0 +1,359 @@
...
+    private fun printLabel(node: Any, label: String) {
+        output.println("\t" + node.hashCode() + " [label=\"" + label + "\"]")
+    }
+
+    override fun visit(node: WorldNode) {
+        var builder = StringBuilder()
+            .append("WorldNode\\nsize = ")
+            .append(node.dimensions.map {
+                when(it.type) {
+                    WorldType.EDGE -> it.size.toString() + "[edge=" + "]"
```
Comment (NicEastvillage) [inline compiler/src/main/kotlin/walkers/grapher.kt]: I think you are missing a the state name in `"[edge=" + "]"`

**Review by NicEastvillage — COMMENTED** [review]
How do I test this?

**Comment by jenrik** [comment]
I'm not sure what you mean by testing it, there isn't really a good way to automatically test this. If you wanna build a graphic from the output you can run the command `dot -Tsvg -o cellmata.svg ast.gs`, and it will produce a svg file

File: compiler/src/main/kotlin/walkers/grapher.kt (hunk 1 above)
Comment (jenrik) [reply compiler/src/main/kotlin/walkers/grapher.kt]: fixed in 398bea9

File: compiler/src/main/kotlin/main.kt [hunk 2]
```diff
@@ -40,6 +42,7 @@ fun compile(path: Path) {
         TypeChecker(symbolTable).visit(ast)
         ErrorLogger.assertNoErrors()
 
+        File("ast.gs").outputStream().use { out -> ASTGrapher(out).visit(ast) }
```
Comment (falkecarlsen) [inline compiler/src/main/kotlin/main.kt:46]: This should be an argument. Like #81, we should use a proper way of parsing arguments.

**Comment by jenrik** [comment]
Here is a online graphviz render https://dreampuf.github.io/GraphvizOnline/

File: compiler/src/main/kotlin/visitors/grapher.kt [hunk 3]
```diff
@@ -0,0 +1,305 @@
...
+
+        if (node.tickrate != null) {
+            builder = builder.append("\\ntickrate=").append(node.tickrate!!)
+        }
+
+        printLabel(node, builder.toString())
+        super.visit(node)
+    }
+
+    override fun visit(node: RootNode) {
```
Comment (falkecarlsen) [inline compiler/src/main/kotlin/visitors/grapher.kt:53]: Could we at least have the RootNode fun override be on top? I don't ask for a completely ordered hierarchy, but some structure could be nice.

File: compiler/src/main/kotlin/visitors/valueextractor.kt [hunk 4]
```diff
@@ -2,6 +2,7 @@ package dk.aau.cs.d409f19.cellumata.visitors
 
 import dk.aau.cs.d409f19.antlr.CellmataParser
 import dk.aau.cs.d409f19.cellumata.ast.*
+import org.antlr.runtime.tree.ParseTree
```
Comment (falkecarlsen) [inline compiler/src/main/kotlin/visitors/valueextractor.kt]: Wait. This doesn't seem right? Why suddenly this import?

**Review by falkecarlsen — CHANGES_REQUESTED** [review]
Minor errors and suggestions. Looks good though, but does not have tons of documentation - this is not strictly necessary.

File: compiler/src/main/kotlin/main.kt (hunk 2 above)
Comment (jenrik) [reply compiler/src/main/kotlin/main.kt:46]: Let's move that change to a separate PR for commandline arguments

File: compiler/src/main/kotlin/visitors/grapher.kt (hunk 3 above)
Comment (jenrik) [reply compiler/src/main/kotlin/visitors/grapher.kt:53]: This seem unimportant

File: compiler/src/main/kotlin/main.kt (hunk 2 above)
Comment (falkecarlsen) [reply compiler/src/main/kotlin/main.kt:46]: Yes, that's fine. Maybe we should open an issue on it.

File: compiler/src/main/kotlin/visitors/grapher.kt (hunk 3 above)
Comment (falkecarlsen) [reply compiler/src/main/kotlin/visitors/grapher.kt:53]: Well yes, but actually no. It's important for readability.

File: compiler/src/main/kotlin/main.kt (hunk 2 above)
Comment (jenrik) [reply compiler/src/main/kotlin/main.kt:46]: I have create a issue, see #86

**Review by falkecarlsen — DISMISSED** [review]

**Review by falkecarlsen — APPROVED** [review]

**Pull request merged**